File: DAY-4/PART-1/code_1.py
def count_word(grid, word):
    rows = len(grid)
    cols = len(grid[0])
    word_len = len(word)
    directions = [
        (0, 1),    # right
        (0, -1),   # left
        (1, 0),    # down
        (-1, 0),   # up
        (1, 1),    # diagonal down right
        (1, -1),   # diagonal down left
        (-1, 1),   # diagonal up right
        (-1, -1)   # diagonal up left
    ]
    count = 0

    def is_valid(x, y):
        return 0 <= x < rows and 0 <= y < cols

    for r in range(rows):
        for c in range(cols):
            for dr, dc in directions:
                match = True
                for k in range(word_len):
                    nr, nc = r + dr * k, c + dc * k
                    if not is_valid(nr, nc) or grid[nr][nc] != word[k]:
                        match = False
                        break
                if match:
                    count += 1

    return count


with open("E:\ADVENT OF CODE\DAY-4\PART-1\code_1.txt", "r") as file:
    grid = [line.strip() for line in file]

result = count_word(grid, "XMAS")

print(f"Answer: {result}")

File: DAY-4/PART-1/code_1.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <regex.h>

#define MAX_BUFFER_SIZE 10000 // Adjust this size based on file size

int main() {
    FILE *file;
    char *filename = "E:\\ADVENT OF CODE\\DAY-3\\PART-1\\code_1.txt";
    char buffer[MAX_BUFFER_SIZE];

    // Open the file
    file = fopen(filename, "r");
    if (file == NULL) {
        printf("Error: Could not open file %s\n", filename);
        return 1;
    }

    // Read the file contents into buffer
    fread(buffer, sizeof(char), MAX_BUFFER_SIZE, file);
    fclose(file);

    // Regular expression for pattern `mul(x, y)`
    const char *pattern = "mul\\(([0-9]+),([0-9]+)\\)";
    regex_t regex;
    regmatch_t matches[3]; // Array to hold match groups (full match + 2 groups)
    int ret;

    // Compile the regex
    ret = regcomp(&regex, pattern, REG_EXTENDED);
    if (ret) {
        printf("Error: Could not compile regex\n");
        return 1;
    }

    char *cursor = buffer;
    int total_sum = 0;

    // Execute the regex
    while (!regexec(&regex, cursor, 3, matches, 0)) {
        // Extract numbers from the match
        char num1[10] = {0}, num2[10] = {0};

        // Get the first number
        int start = matches[1].rm_so;
        int end = matches[1].rm_eo;
        strncpy(num1, cursor + start, end - start);

        // Get the second number
        start = matches[2].rm_so;
        end = matches[2].rm_eo;
        strncpy(num2, cursor + start, end - start);

        // Convert to integers and calculate product
        int x = atoi(num1);
        int y = atoi(num2);
        total_sum += x * y;

        // Move cursor forward to process the next match
        cursor += matches[0].rm_eo;
    }

    // Free the regex memory
    regfree(&regex);

    // Print the result
    printf("Answer: %d\n", total_sum);

    return 0;
}
